utils/pubsub.go

```diff
@@ -14,56 +14,108 @@ type DataEvent struct {
 // DataChannel is a channel which can accept an DataEvent
 type DataChannel chan DataEvent
 
-// DataChannelSlice is a slice of DataChannels
-type DataChannelSlice []DataChannel
 type PublishSubscriber interface {
 	Publish(topic string, data interface{})
-	PublishToChannel(channel DataChannel, topic string, data interface{})
 	Subscribe(topic string, ch DataChannel)
 }
 
 // EventBus stores the information about subscribers interested for a particular topic
 type EventBus struct {
-	subscribers map[string]DataChannelSlice
-	rm          sync.RWMutex
+	lastEventMutex sync.RWMutex
+	// lastEvent holds the last event for each topic so that we can send it to new subscribers
+	lastEvent map[string]*DataEvent
+
+	subscriptionsMutex sync.RWMutex
+	// subscriptions keep the list of subscription publishers per topic
+	subscriptions map[string]subPublishers
 }
 
 func (eb *EventBus) Publish(topic string, data interface{}) {
-	eb.rm.RLock()
-	if chans, found := eb.subscribers[topic]; found {
-		// this is done because the slices refer to same array even though they are passed by value
-		// thus we are creating a new slice with our elements thus preserve locking correctly.
-		// special thanks for /u/freesid who pointed it out
-		channels := append(DataChannelSlice{}, chans...)
-		go func(data DataEvent, dataChannelSlices DataChannelSlice) {
-			for _, ch := range dataChannelSlices {
-				//Publishing to channels in separate goroutines
-				go func(ch DataChannel, data DataEvent) {
-					ch <- data
-				}(ch, data)
-			}
-		}(DataEvent{Data: data, Topic: topic}, channels)
+	eb.subscriptionsMutex.RLock()
+	defer eb.subscriptionsMutex.RUnlock()
+	eb.lastEventMutex.Lock()
+	defer eb.lastEventMutex.Unlock()
+
+	evt := &DataEvent{Data: data, Topic: topic}
+	if eb.lastEvent == nil {
+		eb.lastEvent = map[string]*DataEvent{}
 	}
-	eb.rm.RUnlock()
-}
+	eb.lastEvent[topic] = evt
 
-func (eb *EventBus) PublishToChannel(channel DataChannel, topic string, data interface{}) {
-	eb.rm.RLock()
-	go func() {
-		channel <- DataEvent{Data: data, Topic: topic}
-	}()
-	eb.rm.RUnlock()
+	if subPublishers, found := eb.subscriptions[topic]; found {
+		for _, subPublisher := range subPublishers {
+			subPublisher.publish(evt)
```

Comment:
What is `publish(evt)` doing exactly? Is it doing some I/O or network call? If yes then it is not optimal to block the whole map with the lock while those operations are running, right?

Reply:
publish is a non blocking operation, sets the value and spawns a goroutine.